### PR TITLE
Add missing API functions

### PR DESCRIPTION
**Pull request by erikscheurer**
This PR adds functions that are missing from the [preCICE API](https://github.com/precice/precice/blob/develop/src/precice/SolverInterface.hpp).

This includes:
- Functions for multiple Edges, Triangles, Quads
- `setMeshTetrahedron` and `setMeshTetrahedra`
- `requiresMeshConnectivityFor`, `setMeshAccessRegion`, `getMeshVerticesAndIDs`
- Optional argument `relativ_read_time` for waveform iteration.

The realization of the last point is still up for discussion. At the moment, I have an optional argument as -1.0 and a case distinction in the function itself which I think is not ideal. 
In the [python bindings](https://github.com/precice/python-bindings/blob/02f43d8de0ab87480735975ab8708ecbf00e049b/cyprecice/cyprecice.pyx#L879-L942) the optional argument is `None`, however, in Julia, if you do this, you can't give the argument Type as `Float64` because `nothing` is of type `Nothing`.
Another option is to simply use multiple dispatch for the function, so create multiple functions just like in the API, however, I did not get this to work at the moment. Trying to run the solverdummies, they could not find the correct function when the functioncall is separated.

### CODE REVIEW

File: src/PreCICE.jl [hunk 1]
```diff
@@ -508,6 +532,39 @@ function setMeshEdge(meshName::String, firstVertexID::Integer, secondVertexID::I
...
+```julia-repl
+julia> vertices = [1 2; 1 3; 1 4; 2 3; 2 4; 3 4]
+julia> vertices.shape
+(6,2)
+julia> setMeshEdges("MeshOne", vertices)
+```
+"""
+function setMeshEdges(meshName::String, vertices::AbstractArray{Cint})
+    _size, dimensions = size(vertices)
+    @assert dimensions == 2 "Dimensions of vector data in write_vector_data does not match with dimensions in problem definition. Provided dimensions: $dimensions, expected dimensions: $(getDimensions())"
```
Comment (IshaanDesai) [inline src/PreCICE.jl:557]: Assertion message seems like a left over from copy pasting. Lets write a more sensible message here. For example, *Vertices pairs need to be provided while setting mesh edges. Please check the shape of the vertices array*.

File: src/PreCICE.jl [hunk 2]
```diff
@@ -554,15 +611,26 @@ Set mesh triangle from vertex IDs.
...
+```julia-repl
+julia> vertices = [1 2 3; 1 3 4; 1 2 4; 1 3 4]
+julia> vertices.shape
+(4,3)
+julia> setMeshTriangles("MeshOne", vertices)
+```
 """
 function setMeshTriangles(meshName::String, vertices::AbstractArray{Integer})
+    _size, dimensions = size(vertices)
+    @assert dimensions == 3 "Dimensions of vector data in write_vector_data does not match with dimensions in problem definition. Provided dimensions: $dimensions, expected dimensions: $(getDimensions())"
```
Comment (IshaanDesai) [inline src/PreCICE.jl:625]: Same as above.

File: src/PreCICE.jl [hunk 3]
```diff
@@ -618,18 +686,82 @@ Set mesh Quad from vertex IDs.
 - `vertices::AbstractArray{Integer}`: IDs of the edges of the Quads.
 
 """
-function setMeshQuad(meshName::String, vertices::AbstractArray{Integer})
+function setMeshQuads(meshName::String, vertices::AbstractArray{Integer})
+    _size, dimensions = size(vertices)
+    @assert dimensions == 4 "Dimensions of vector data in write_vector_data does not match with dimensions in problem definition. Provided dimensions: $dimensions, expected dimensions: $(getDimensions())"
```
Comment (IshaanDesai) [inline src/PreCICE.jl:691]: Same as the previous comment for edges.

File: src/PreCICE.jl [hunk 4]
```diff
@@ -618,18 +686,82 @@ Set mesh Quad from vertex IDs.
...
+Set mesh Tetrahedron from vertex IDs.
+
+# Arguments
+- `meshName::String`: Name of the mesh to add the Tetrahedron to.
+- `vertices::AbstractArray{Integer}`: IDs of the edges of the Tetrahedra.
+
+"""
+function setMeshTetrahedra(meshName::String, vertices::AbstractArray{Integer})
+    _size, dimensions = size(vertices)
+    @assert dimensions == 4 "Dimensions of vector data in write_vector_data does not match with dimensions in problem definition. Provided dimensions: $dimensions, expected dimensions: $(getDimensions())"
```
Comment (IshaanDesai) [inline src/PreCICE.jl:754]: Same as above. Also lets find a better name for the variable than **dimensions**.

File: src/PreCICE.jl [hunk 5]
```diff
@@ -836,9 +968,9 @@ end
 
 @doc """
 
-    readBlockVectorData(meshName::String, dataName::String, valueIndices::AbstractArray{Cint})::AbstractArray{Float64}
+    readBlockVectorData(meshName::String, dataName::String, valueIndices::AbstractArray{Cint}[, relative_read_time::Float64])
 
-Read and return vector data values given as block.
+Read and return vector data values given as block.
```
Comment (IshaanDesai) [inline src/PreCICE.jl:973]:
```suggestion
Read and return vector data values given as block.
```

File: src/PreCICE.jl [hunk 6]
```diff
@@ -875,32 +1008,50 @@ function readBlockVectorData(
...
+            meshName,
+            dataName,
+            _size,
+            valueIndices,
+            values,
+            relative_read_time,
+        )
+    end
+
     return permutedims(reshape(values, (getDimensions(), _size)))
```
Comment (IshaanDesai) [inline src/PreCICE.jl:1041]: Why do we return something for a read operation?

File: src/PreCICE.jl [hunk 7]
```diff
@@ -875,32 +1008,50 @@ function readBlockVectorData(
...
 end
 
 @doc """
 
-    readVectorData(meshName::String, dataName::String, valueIndex::Integer)::AbstractArray{Float64}
+    readVectorData(meshName::String, dataName::String, valueIndex::Integer[, relative_read_time::Float64])::AbstractArray{Float64}
```
Comment (IshaanDesai) [inline src/PreCICE.jl:1046]: Same concern as above.

**Review by IshaanDesai — CHANGES_REQUESTED** [review]
Looks good already :+1: some comments from my side.

File: src/PreCICE.jl [hunk 8]
```diff
@@ -836,9 +968,9 @@ end
 
 @doc """
 
-    readBlockVectorData(meshName::String, dataName::String, valueIndices::AbstractArray{Cint})::AbstractArray{Float64}
+    readBlockVectorData(meshName::String, dataName::String, valueIndices::AbstractArray{Cint}[, relative_read_time::Float64])
```
[the remark this answers is not in the document]
Comment (erikscheurer) [reply src/PreCICE.jl:971]: Yes, as long as there is no default value (which in our case semantically is not) this is the [correct](https://docs.julialang.org/en/v1/manual/documentation/#Writing-Documentation) syntax.

File: .github/workflows/build-and-test.yml [hunk 9]
```diff
@@ -15,7 +15,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        julia-version: ["1.6.0", "1.6.5", "1.7.0", "^1.7"]
+        julia-version: ["1.6.7", "1.7", "^1.8"]
```
[the remark this answers is not in the document]
Comment (IshaanDesai) [reply .github/workflows/build-and-test.yml:18]: Okay got it :+1: sounds very reasonable.

**Review by IshaanDesai — APPROVED** [review]

**Pull request merged**